DOCUMENTATION.md

```diff
@@ -43,6 +41,314 @@ The overall territory is laid out here in this [C4](https://c4model.com) System
+- Anything that fails can be deferred for later retry pretty naturally (i.e. set a visibility timeout on the item)
+- Any case which for some reason triggers an exception, can be dealt with as a Poison Message
+- After a configured amount of time or attempts, you can have messages shift to a Dead Letter Queue. There are established practices regarding alerting in such circumstances and/or triggering retries of Dead Lettered messages  
+- Arranging processing via a Queue also gives a natural way for other systems (or ad-hoc admin tools) to trigger such messages too. For example, you can easily do a manual trigger or re-trigger of some reaction by adding a message out of band.
+- Queues are a generally applicable technique in a wide variety of systems, and look great on a system architecture diagram or a resume
+
+There are also a number of potential negatives; the significance of that depends on your needs:
+- You have another system to understand, worry about the uptime of, and learn to monitor
+- Marking a message invisible (or running competing consumers) will mean you can receive messages regarding a particular action out of order. In other words, you get no useful ordering guarantees (frequently this is not a problem)
+- If the time taken or processing cost of the messages on the Queue are not uniform, having 10,000 or a million messages in the queue might be hard to interpret in terms of how long it will take to clear. You also have no way to know where your high value orders are relative to that batch of 400,000 failures from that third party system 10 minute service outage that will retry in an hours time
+- there are no great ways to 'rewind the position' and retraverse events over a particular period of time without causing lots of confusion
+- running a newer version alongside an existing one (to Expand then Contract for a Read Model) etc involves making a new Queue instance, making process that writes the messages also write to the Queue instance, validating that both queues always get the same messages etc. (As opposed to working direct off the Store notifications - you can just use a new Consumer Group Name, seeded with a nominated Position)
+
+Many of the pleasing properties are not (or at least not directly) provided by a Reaction processing engine like Propulsion, but there are alternate approaches:
+- If the processor hits a poison message, checkpointing cannot progress beyond the Batch the triggering event was in. Processing will continue until the configured maximum batches read ahead count has been reached, but will then stop. This can be mitigated by alerting  based on the built in metrics that are provided regarding the maximum time any stream has been processing ('stuck' messages), or failing (the typical 'poison' message case). (Arguably that's better than risking 40,000 clones of that same poison message over the next 12 hours, affecting all high value orders).
```

Comment:
I'm struggling with this point. How does alerting mitigate these situations (alerting by definition happens after this has become a problem).

Suggesting a re-word mostly to see if we're thinking the same thing

```suggestion
- Propulsion allows you to configure a maximum number of batches to read ahead in case of a poison message. Checkpointing will progress beyond the Batch containing the poison event. Processing will continue until the configured maximum batches read ahead count has been reached, but will then stop. It is important to be aware of these scenarios and mitigate their impact by catching them early. Alerting should be set up based on the built in metrics that are provided regarding the maximum time any stream has been processing ('stuck' messages), or failing (the typical 'poison' message case). Knowing this scenario is happening is the first step in recovering.
```

Reply:
Depending on how long messages take to process and SLA, knowing something has been stuck or failing for 2 minutes might trigger a low level alert and 15 mins a proper one
You wouldnt do that with a queue, but for a notification pipeline, you can observe the live stuck/failing time
Will try to reword as this is far too subtle to try to convey in some lazily written long sentence as I did

High level: you write stuff to not fail, and, assuming continuous delivery, tend to fix problems by code deploys (e.g. skipping a poison message). Being able to do that sort of thing relies on reliable alerting - fail counts etc is not enough.
Dont have a great example either...

Reply:
Might be easier to explain it from the opposite angle: "The whole point of a queue is to delay processing of items to normalise spikes in traffic. If something is time sensitive a queue is likely not the right place for it"

But I think that points already been made 😄

Reply:
This prompted a significant rewrite of stuff touching poison messages in https://github.com/jet/propulsion/pull/200/commits/9dd0d9cdb7aba540b3a4e596af09dc84b3d9d629
Comments welcome, but probably best after I address [the priority queues stuff comment](https://github.com/jet/propulsion/pull/200#discussion_r1097351507)

Reply:
Added more info about how to manage/design/alert on stuck projectors https://github.com/jet/propulsion/pull/200/commits/be770f1b11f903f6f2e662a67276822c257eb44f
reading link https://github.com/jet/propulsion/blob/doc/DOCUMENTATION.md#busy-metrics